sdk/client/src/verifying_client.rs

```diff
@@ -0,0 +1,516 @@
+        requests.push(MethodRequest::get_state_proof(start_version));
+
+        // actually send the batch
+        let responses = self.inner.batch(requests).await?;
+
+        // TODO(philiphayes): REMOVE. for now, just convert the errors until we
+        // use the real error type.
+        let mut responses = responses
+            .into_iter()
+            .map(|result| result.map_err(anyhow::Error::new))
+            .collect::<Vec<_>>();
+
+        // extract and verify the state proof
+        let (state_proof_response, state) = responses.pop().unwrap()?.into_parts();
+        let state_proof_view = state_proof_response.try_into_get_state_proof()?;
+        let state_proof: StateProof = (&state_proof_view).try_into()?;
+
+        // check the response metadata matches the state proof
+        let latest_li = state_proof.0.ledger_info();
+        ensure!(state.version == latest_li.version());
+        ensure!(state.timestamp_usecs == latest_li.timestamp_usecs());
+
+        // try to ratchet our trusted state using the state proof
+        self.verify_and_ratchet(&state_proof)?;
+
+        // remote says we're too far behind and need to sync. we have to throw
+        // out the batch since we can't verify any proofs
+        if state_proof.1.more {
+            bail!("needs sync");
+        }
+
+        // unflatten subresponses, verify, and collect into one response per request
+        batch.validate_responses(start_version, &state, &state_proof, responses)
+    }
+}
+
+struct VerifyingBatch {
+    requests: Vec<VerifyingRequest>,
+}
+
+impl VerifyingBatch {
+    fn from_batch(requests: Vec<MethodRequest>) -> Self {
+        Self {
+            requests: requests.into_iter().map(VerifyingRequest::from).collect(),
+        }
+    }
+
+    fn num_subrequests(&self) -> usize {
+        self.requests
+            .iter()
+            .map(|request| request.subrequests.len())
+            .sum()
+    }
+
+    fn collect_requests(&self) -> Vec<MethodRequest> {
+        self.requests
+            .iter()
+            .flat_map(|request| request.subrequests.iter().cloned())
+            .collect()
+    }
+
+    fn validate_responses(
+        &self,
+        start_version: Version,
+        state: &State,
+        state_proof: &StateProof,
+        responses: Vec<Result<Response<MethodResponse>>>,
+    ) -> Result<Vec<Result<Response<MethodResponse>>>> {
+        ensure!(
+            self.num_subrequests() == responses.len(),
+            "unexpected number of responses"
+        );
+
+        for response in &responses {
+            if let Ok(response) = response {
+                ensure!(response.state() == state);
+            }
+        }
+
+        let mut responses_iter = responses.into_iter();
+
+        Ok(self
+            .requests
+            .iter()
+            .map(|request| {
+                let n = request.subrequests.len();
+                let subresponses = responses_iter.by_ref().take(n).collect();
```

Comment:
it is possible request response is not in first nth elements, should use id to find response

Reply:
See my other comment on the rust sdk batching behavior